### PR TITLE
Verify sender's balance before importing transaction to queue

### PR DESCRIPTION
**Pull request by tomusdrw**
+ Additional trace logging

### CODE REVIEW

**Comment by arkpar** [comment]
Test fails

**Comment by tomusdrw** [comment]
Yes, sorry. Fixing it (after changing validation to include full transaction cost balance in test was insufficient).

**Comment by arkpar** [comment]
Could you please check that this does not affect the block propagation logic? It should not send out any blocks (and transactions) while the sync is still active, and I believe this was done by not sending the chain changed event until the sync is complete.

**Comment by gavofyork** [comment]
ditto. also, right now the client and sync logic is extremely brittle. please place plentiful comments/documentation around any logic changed.

File: rpc/src/v1/impls/eth.rs [hunk 1]
```diff
@@ -370,7 +370,10 @@ impl<C, S, A, M, EM> Eth for EthClient<C, S, A, M, EM>
 						let signed_transaction = transaction.sign(&secret);
 						let hash = signed_transaction.hash();
 
-						let import = miner.import_transactions(vec![signed_transaction], |a: &Address| client.nonce(a));
+						let import = miner.import_transactions(vec![signed_transaction], |a: &Address| AccountDetails {
+							nonce: client.nonce(a),
+							balance: client.balance(a)
```
Comment (gavofyork) [inline rpc/src/v1/impls/eth.rs]: missing comma.

File: sync/src/chain.rs [hunk 2]
```diff
@@ -946,8 +946,11 @@ impl ChainSync {
 			transactions.push(tx);
 		}
 		let chain = io.chain();
-		let fetch_nonce = |a: &Address| chain.nonce(a);
-		let _ = self.miner.import_transactions(transactions, fetch_nonce);
+		let fetch_account = |a: &Address| AccountDetails {
+			nonce: chain.nonce(a),
+			balance: chain.balance(a)
```
Comment (gavofyork) [inline sync/src/chain.rs]: missing comma

**Comment by tomusdrw** [comment]
After chat with @arkpar I've removed the change in client (it should be in separate pr anyway).

**Comment by gavofyork** [comment]
ok - lgtm then.

**Pull request merged**